### PR TITLE
Improve handling of zero range datashader aggregations

### PR DESCRIPTION
**Pull request by philippjfr**
Ensures datashader aggregation on zero-range data is handled correctly. Previously there were some workarounds that did one of two things:

1. If the range was zero-width then it was padded by an arbitrary factor of 0.5
2. If the width or height were 0 then it would simply increase it to 1

Both were only workarounds and led to unexpected behavior so the operations now return Images with zero sized arrays. This is also related to https://github.com/bokeh/datashader/pull/612 but doesn't require it since the operations are doing all the guarding against the bad cases that datashader doesn't handle correctly.

- [x] Fixes https://github.com/ioam/holoviews/issues/2817
- [x] Adds unit tests

### CODE REVIEW

**Comment by jbednar** [comment]
Looks good, thanks.

**Comment by jbednar** [comment]
New changes look good too.

**Comment by philippjfr** [comment]
Ready to review.

**Comment by jbednar** [comment]
Looks good to me, but would be helpful if @jlstevens looks it over too.

**Comment by philippjfr** [comment]
One arbitrary thing I've done here is set the density for zero sized dimensions to 1, technically the density should be ``np.inf`` or something but that causes all kinds of trouble.

File: holoviews/operation/datashader.py [hunk 1]
```diff
@@ -468,6 +458,27 @@ def _process(self, element, key=None):
             vdims = Dimension('Count')
         params['vdims'] = vdims
 
+        if x is None or y is None or width == 0 or height == 0:
+            xarray = xr.DataArray(np.full((height, width), np.NaN, dtype=np.float32),
```
Comment (jlstevens) [inline holoviews/operation/datashader.py]: Why does it have to be ``np.float32`` in these two cases?

Comment (philippjfr) [reply holoviews/operation/datashader.py]: No good reason, pointless optimization. Can take it out.

Comment (jlstevens) [reply holoviews/operation/datashader.py]: I wouldn't specify a precision unless there is a good reason.

**Comment by jlstevens** [comment]
Looks good to me and the tests are passing. Merging.

**Pull request merged**